fix(snap): defer service startup till after configure hook runs

tonyespy opened the pull request:
This PR modifies the way the snap install and configure hooks when the snap in installed in order to ensure that snap configuration provided at install time is processed before services are started. This is working around an issue with snapd where snap configuration at install time is not made available to the snap until the configure hook is called, which is after the services have already been started.

So, this PR disables all of the services in the install hook, and then in the configure hook, the services are started after configuration has been processed.

<!-- Expected Commit Message Description (imported automatically by GitHub) -->
<!-- Must conform to [conventional commits guidelines](https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md) -->
<!-- Expected Commit message must contain Closes/Fixes #IssueNumber statement when there is a related issue -->

<!-- Add additional detailed description of need for change if no related issue -->

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [*] I have added unit tests for the new feature or bug fix (if not, why?)
  - units tests to be added before this is taken out of draft status 
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [NA] I have opened a PR for the related docs change (if not, why?)
  - This fix relates to the install behavior of the snap, and as such should be transparent to the installer. 

## Testing Instructions
 - Install the snap built from this PR and verify that all of the expected services start
 - Confirm that device-virtual can be enabled (`$ sudo snap set edgexfoundry device-virtual=on`)
 - Confirm that kuiper can be enabled (`$ sudo snap set edgexfoundry kuiper=on`); ensure that both kuiper and app-service-configurable are started
 - Confirm that support-notifications and support-scheduler can be started (e.g. `snap set... support-scheduler=on`)
 - Confirm that security-proxy can be disabled (`$sudo snap set security-proxy=off`)

Additional test instructions will be added to this PR as it progresses. The more advanced cases require a custom gadget snap which provides default configuration. This is currently the only way to disable secret-store usage. If possible, this PR will be extended to support dynamically disabling secret store usage, but this is currently blocked on issue #3712.

Also note, this has been tested using [our in-progress checkbox provider for Jakarta](https://code.launchpad.net/~mengyiw/checkbox-provider-edgex/+git/checkbox-provider-edgex/+merge/409927) and all of the tests pass.

## New Dependency Instructions (If applicable)
<!-- Please follow [vetting instructions](https://wiki.edgexfoundry.org/display/FA/Vetting+Process+for+3rd+Party+Dependencies) and place results here -->

tonyespy commented:
FYI, just force pushed a minor update to address a minor gofmt error which was tripping up the CI build.

siggiskulason reviewed (commented):
Looks good so far. I followed the instructions and could start the services, stop the proxy and start it up again and all the services restarted correctly.

anonymouse64 commented:
Just a general comment that the change to remove the device-virtual service seems to be also deleted but it isn't mentioned in the commit messages or the PR description.

anonymouse64 commented on snap/local/hooks/cmd/configure/configure.go:
```diff
@@ -86,146 +109,262 @@ func serviceType(name string) int {
 }
 
+func appendServices(startServices []string, newServices []string) []string {
+	for _, s := range newServices {
+		startServices = append(startServices, s)
+	}
+	return startServices
```
this can also just be simplified to one append:

```suggestion
	return append(startServices, newServices...)
```

anonymouse64 commented on snap/local/hooks/cmd/configure/configure.go:
```diff
@@ -86,146 +109,262 @@ func serviceType(name string) int {
 // been started, the service must be restarted manually.
 //
-func handleAllServices() error {
+func handleAllServices(install bool) (error, []string) {
```
a comment about the install parameter would be nice

anonymouse64 commented on snap/local/hooks/cmd/configure/configure.go:
```diff
@@ -86,146 +109,262 @@ func serviceType(name string) int {
+		// NOTE -- see addtl comments about *always* required services.
+		// These should really be removed from the set of configurable
+		// services (see hooks.Services) as if they were actually disabled
+		// the syste would no longer work. If/when they're removed, then
```
```suggestion
		// the system would no longer work. If/when they're removed, then
```

anonymouse64 commented on snap/local/hooks/cmd/configure/configure.go:
```diff
@@ -86,146 +109,262 @@ func serviceType(name string) int {
+			case ON:
+				serviceList = getKuiperServices()
+				if install {
+					startServices = appendServices(startServices, serviceList)
```
this helper can go away completely if you do

```suggestion
					startServices = append(startServices, serviceList...)
```

anonymouse64 replied:
I guess even better you could do

```suggestion
					startServices = append(startServices, getKuiperServices()...)
```

anonymouse64 replied:
ah nevermind I see you do also need serviceList too, so ignore the second recommendation

anonymouse64 commented on snap/local/hooks/cmd/configure/configure.go:
```diff
@@ -86,146 +109,262 @@ func serviceType(name string) int {
-					return err
+				serviceList = getProxyServices()
+				if install {
+					startServices = appendServices(startServices, serviceList)
```
same

anonymouse64 commented on snap/local/hooks/cmd/install/install.go:
```diff
@@ -389,22 +426,48 @@ func main() {
 		}
 	}
+
+	if err = cli.SetConfig("install", "true"); err != nil {
```
it does seem inconsistent to me in hindsight that snapd let's you set config items in the install hook (before the configure hook runs), but it doesn't let you read them before the configure hook runs :-|

anonymouse64 commented on snap/local/hooks/cmd/install/install.go:
```diff
@@ -323,7 +348,19 @@ func installProxy() error {
 
 func installRedis() error {
-	if err := os.MkdirAll(hooks.SnapData+"/redis", 0755); err != nil {
+	setupScriptPath, err := exec.LookPath("install-setup-redis.sh")
```
minor nitpick, but this whole file could be implemented in Go with less lines of code than running the script here:

```go
snapData := os.Getenv("SNAP_DATA")
fileName := filepath.Join(snapData,"/redis/conf/redis.conf")
if _, err := os.Stat(filepath.Join(snapData,"redis")); err != nil {
    // dir doesn't exist
    if err := os.MkdirAll(filepath.Dir(fileName),0755); err != nil {
        return err
    }
    if err := ioutil.WriteFile(fileName, nil, 0644); err != nil {
         return err
    }
}
```

with better comments of course

anonymouse64 commented on snap/local/hooks/cmd/configure/configure.go:
```diff
@@ -250,8 +391,33 @@ func main() {
+		installMode = true
+	}
+
+	if err, startServices = handleAllServices(installMode); err != nil {
```
a comment that handleAllServices will dynamically stop/start services if installMode is false would be nice

anonymouse64 replied:
otherwise it seems that handleAllServices just returns the set of services which should be started, but really it only returns the list of services that should be started when we are in install mode

anonymouse64 commented on snap/local/hooks/cmd/configure/configure.go:
```diff
@@ -86,146 +109,262 @@ func serviceType(name string) int {
 		status, err := cli.Config(s)
 		if err != nil {
-			return err
+			return err, startServices
```
you can just do `return err, nil` everywhere, that is more Go idiomatic

anonymouse64 reviewed (commented):
some comments, overall lgtm

tonyespy commented:
> Just a general comment that the change to remove the device-virtual service seems to be also deleted but it isn't mentioned in the commit messages or the PR description.

@anonymouse64 First, thanks for review!

Re: your device-virtual comment, if you look again, you'll notice that what I removed was an actual case statement for device-virtual. That's because the original case for device-virtual in `handleAllServices` had additional logic to re-enable core-data and core-metadata when device-virtual is enabled.  In the Ireland (2.0) release of EdgeX, device services now publish events direct to the message bus by default, so core-data is no longer a hard-dependency. Regarding core-metadata, if it's not running the whole system falls apart, so we should really remove the ability to disable it, but I decided to leave that for another day. Since there are no dependencies to handle, device-virtual is simply handed like any other service without dependencies (i.e. by the default case).

farshidtz commented:
You can upgrade the edgex-snap-hooks to use the newly added function. See https://github.com/canonical/edgex-snap-hooks/releases/tag/v2.1.0

```bash
go get github.com/canonical/edgex-snap-hooks/v2@v2.1.0
```

I've explained [here](https://github.com/canonical/edgex-snap-hooks/pull/18#issuecomment-962109016) that the [EnabledServices](https://pkg.go.dev/github.com/canonical/edgex-snap-hooks/v2#CtlCli.EnabledServices) function is not useful when called in the install hook.

tonyespy commented:
I just force-pushed a rebase which uses the new edgex-snap-hooks version @farshidtz recently published. This uses the new `UnsetConfig` and `StartMultiple` methods. I've re-run the checkbox tests against the snap, and all continue to pass.

tonyespy commented:
Just force-pushed a rebase to update `go.mod/go.sum` to edgex-snap-hooks 2.1.1 to pickup @siggiskulason's fix for TLS configuration.

tonyespy commented:
Just pushed a new commit which enables the Secret Store to be disabled at runtime. Thanks again @farshidtz for figuring out issue #3712.

tonyespy commented:
Closing as this PR has now been superseded by PR #3813.
